tensorflow/core/kernels/sparse_fill_empty_rows_op_gpu.cu.cc

```diff
@@ -0,0 +1,674 @@
+/* Copyright 2021 The TensorFlow Authors. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+#if GOOGLE_CUDA || TENSORFLOW_USE_ROCM
+
+#define EIGEN_USE_GPU
+
+#include "tensorflow/core/common_runtime/gpu/gpu_event_mgr.h"
+#include "tensorflow/core/framework/register_types.h"
+#include "tensorflow/core/framework/tensor_types.h"
+#include "tensorflow/core/kernels/gpu_prim.h"
+#include "tensorflow/core/kernels/sparse_fill_empty_rows_op.h"
+#include "tensorflow/core/platform/types.h"
+#include "tensorflow/core/util/gpu_device_functions.h"
+#include "tensorflow/core/util/gpu_kernel_helper.h"
+#include "third_party/eigen3/unsupported/Eigen/CXX11/Tensor"
+
+#if GOOGLE_CUDA
+#include "tensorflow/core/util/cuda_solvers.h"  // For ScratchSpace
+#include "tensorflow/stream_executor/cuda/cuda_activation.h"
+using stream_executor::cuda::ScopedActivateExecutorContext;
+#elif TENSORFLOW_USE_ROCM
+#include "tensorflow/core/util/rocm_solvers.h"
+#include "tensorflow/stream_executor/rocm/rocm_activation.h"
+using stream_executor::rocm::ScopedActivateExecutorContext;
+#endif
+
+namespace tensorflow {
+
+typedef Eigen::GpuDevice GPUDevice;
+
+namespace kernel_forward {
+bool to_pointers(bool x) { return x; }
+int32 to_pointers(int32 x) { return x; }
+int64 to_pointers(int64 x) { return x; }
+template <class T>
+T* to_pointers(T* x) {
+  return x;
+}
+template <class T>
+typename T::PointerType to_pointers(T& x) {
+  return x.data();
+}
+template <class T>
+typename T::ConstPointerType to_pointers(const T& x) {
+  return x.data();
+}
+
+template <typename Tindex, typename... CallerArgs, typename... KernelArgs>
+Status wrap_kernel_call(void (*func)(KernelArgs...), const GPUDevice& device,
+                        Tindex size, CallerArgs... args) {
+  auto config = GetGpuLaunchConfig(size, device);
+  return GpuLaunchKernel(func, config.block_count, config.thread_per_block, 0,
+                         device.stream(), config, to_pointers(args)...);
+}
+};  // namespace kernel_forward
+
+using kernel_forward::wrap_kernel_call;
+
+namespace functor {
+
+namespace {
+
+// Computes elements_per_row[0..dense_rows] and sets *rows_are_not_ordered to
+// true if the indices are not ordered by row.
+template <typename Tindex>
+__global__ __launch_bounds__(1024) void CountElementsPerRowKernel(
+    GpuLaunchConfig cfg, Tindex dense_rows, int rank, const Tindex* indices,
```

Comment:
Is `rank` always `2`?

Reply:
No, it can be higher for higher-rank sparse tensors.